dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp

```diff
@@ -408,318 +407,114 @@ void DAGQueryBlockInterpreter::executeCastAfterTableScan(
-    std::unordered_set<String> column_set_for_other_join_filter;
-    bool make_nullable = join.join_type() == tipb::JoinType::TypeRightOuterJoin;
-    for (auto const & p : input_streams_vec[0][0]->getHeader().getNamesAndTypesList())
-    {
-        join_output_columns.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
-        columns_for_other_join_filter.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
-        column_set_for_other_join_filter.emplace(p.name);
-    }
-    make_nullable = join.join_type() == tipb::JoinType::TypeLeftOuterJoin;
-    for (auto const & p : input_streams_vec[1][0]->getHeader().getNamesAndTypesList())
-    {
-        if (!is_semi_join)
-            /// for semi join, the columns from right table will be ignored
-            join_output_columns.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
-        /// however, when compiling join's other condition, we still need the columns from right table
-        columns_for_other_join_filter.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
-        column_set_for_other_join_filter.emplace(p.name);
-    }
+    JoinInterpreterHelper::TiflashJoin tiflash_join{join};
 
-    bool is_tiflash_left_join = kind == ASTTableJoin::Kind::Left || kind == ASTTableJoin::Kind::Cross_Left;
-    /// Cross_Right join will be converted to Cross_Left join, so no need to check Cross_Right
-    bool is_tiflash_right_join = kind == ASTTableJoin::Kind::Right;
-    /// all the columns from right table should be added after join, even for the join key
-    NamesAndTypesList columns_added_by_join;
-    make_nullable = is_tiflash_left_join;
-    for (auto const & p : right_pipeline.streams[0]->getHeader().getNamesAndTypesList())
-    {
-        columns_added_by_join.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
-    }
+    DAGPipeline probe_pipeline;
+    DAGPipeline build_pipeline;
+    probe_pipeline.streams = input_streams_vec[1 - tiflash_join.build_side_index];
+    build_pipeline.streams = input_streams_vec[tiflash_join.build_side_index];
 
-    String match_helper_name;
-    if (is_left_semi_family)
-    {
-        const auto & left_block = input_streams_vec[0][0]->getHeader();
-        const auto & right_block = input_streams_vec[1][0]->getHeader();
+    RUNTIME_ASSERT(!input_streams_vec[0].empty(), log, "left input streams cannot be empty");
+    const Block & left_input_header = input_streams_vec[0].back()->getHeader();
 
-        match_helper_name = Join::match_helper_prefix;
-        for (int i = 1; left_block.has(match_helper_name) || right_block.has(match_helper_name); ++i)
-        {
-            match_helper_name = Join::match_helper_prefix + std::to_string(i);
-        }
+    RUNTIME_ASSERT(!input_streams_vec[1].empty(), log, "right input streams cannot be empty");
+    const Block & right_input_header = input_streams_vec[1].back()->getHeader();
```

Comment:
build/probe is the concept of execution.
Regardless of who is the build side, the order of the output blocks is left -> right.

Reply:
So I won't replace all left/right by probe/build